### PR TITLE
Feat/add connectivity methods

### PR DESCRIPTION
**Pull request by beroda**
### Overview
This should resolve https://github.com/pyvista/pyvista/issues/3598

### CODE REVIEW

**Comment by beroda** [comment]
@banesullivan Following your comment https://github.com/pyvista/pyvista/issues/3598#issuecomment-1314604748 , I successfully tested these developments  against various dataset types:
- pyvista.core.pointset.UnstructuredGrid
- pyvista.core.pointset.PolyData
- pyvista.core.grid.RectilinearGrid
- pyvista.core.pointset.StructuredGrid
-  pyvista.core.grid.UniformGrid

The only problem I encountered is related to **pyvista.core.pointset.ExplicitStructuredGrid**.
I don't know if it comes from a vtk bug or from my implementations. I might need some help on that one.

@MatthewFlamm FYI

**Comment by beroda** [comment]
Hello everyone. 

Could someone give me a hand on this PR ?
It is not finished yet but I encountered a weird mypy error.

Also, I would highly appreciate some feedback on the implementations, that would be great. Probably @banesullivan or @MatthewFlamm since you commented on the original issue ? I added some examples in the docstrings. Thank you in advance :)

**Comment by MatthewFlamm** [comment]
I have been very busy on other things, but would like to review this if possible.  I've overarching comment without reading in detail first.  This is for discussion and I'm not asking for changes unless others agree.

This PR implements methods related to cell connectivity, but it deviates in structure from vtk.  vtkCell has more to it, for example querying where a point in space is inside a cell. See https://vtk.org/doc/nightly/html/classvtkCell.html#a188d8c01ab1cc03e1fd8b04e3ec97faf.   Should we expand this new class to be inclusive of this kind of usage?  I'm not suggesting to do it all in this PR, but align the structure and naming going forward to enable this.

**Comment by banesullivan** [comment]
Sorry, I too am quite busy lately. +1 to @MatthewFlamm's comment about structure and deviating from VTK. I want to be *very careful* adding a new class like this that deviates from having a VTK counterpart. If we could move this closer to wrapping `vtkCell` that would be better, but I need more time to look into this to provide any clear direction for you

**Comment by MatthewFlamm** [comment]
Based on the additional comment, what makes sense to me is to wrap vtkCell and then allow accessing it via either `dataset.cells[0]` or `dataset.cells(0)`.  I would lean to the former usage although in vtk it behaves more like the latter.  Then we could do things like:

```py
dataset.cells[0].point_ids
```


There may be some subtleties around creating cells that would need to be thought through, but we don't necessarily have to allow that in the first implementation.

**Comment by beroda** [comment]
> Based on the additional comment, what makes sense to me is to wrap vtkCell and then allow accessing it via either `dataset.cells[0]` or `dataset.cells(0)`. I would lean to the former usage although in vtk it behaves more like the latter. Then we could do things like:
> 
> ```python
> dataset.cells[0].point_ids
> ```

The only problem I can think of is that if we only wrap **vtkCell**, we loose the point to cell connectivity.
Maybe we can take the best of both worlds and use the not-already-wrapped **vtkCell** in the class I implemented.
That can be done very easily I think.

Please, feel free to modify this PR :)

**Comment by MatthewFlamm** [comment]
On reading this again, I think my first impression was not entirely correct, although I think my comment how this interfaces with a wrapping of `vtkCell` still needs to be considered.  Sorry for the confusion.  I see that this PR also relies heavily on methods that come from the `DataSet`s like `GetCellNeighbors`, which would not be addressed by wrapping `vtkCell` as a first rate PyVista class.

**Comment by akaszynski** [comment]
### Proposal
Looking at this with fresh eyes. @beroda, since #3715 effectively exposes much of the functionality of `DatasetConnectivity`, I recommend that instead of:

```py
mesh.connectivity_cells.get_neighbors(..., connections="cells")
```

we instead implement:
```py
mesh.cell_neighbors()
```

### Reasoning
This closely matches the existing VTK API [DataSet.GetCellNeighbors](https://vtk.org/doc/nightly/html/classvtkDataSet.html#ae1ba413c15802ef50d9b1955a66521e4), and all other APIs (like `_cell_point_ids`, `_point_cell_ids`, and `_cell_neighbors` are covered by `pyvista.Cell` accessible from `pyvista.DataSet.get_cell`.

This allows us to align with VTK's existing API and matches the Python  ["Flat is better than nested"](https://medium.com/@ankushchoubey/clean-code-1-flat-is-better-than-nested-leave-when-not-okay-c09ba74090ef), which can be aptly applied here.

### Considerations

We could also implement the following:
- [GetPointCells](https://vtk.org/doc/nightly/html/classvtkDataSet.html#a36d1d8f67ad67adf4d1a9cfb30dade49)
- [FindCell](https://vtk.org/doc/nightly/html/classvtkDataSet.html#a2221c10d3c4cca44e82c5ef70e4e1cbd)
- [GetCellPoints](https://vtk.org/doc/nightly/html/classvtkDataSet.html#ab4d677c257a58e4eb4a80757f9b371ea)

**Comment by beroda** [comment]
@akaszynski You are right, looking at what has been implemented in https://github.com/pyvista/pyvista/pull/3715, I think the `DataSetConnectivity` class is no longer needed an can be split into `Dataset'`s methods.

I also agree that your proposal will look more familiar to VTK's API and will help users understand what's behind the scenes.

If it is fine for you, I can ping you when the PR is ready so that you can have a look (and add Pink Floyd quotes :wink:)

**Comment by akaszynski** [comment]
> If it is fine for you, I can ping you when the PR is ready

Perfect. I look forward to reviewing it.

**Comment by akaszynski** [comment]
@beroda, is this ready for review?

**Comment by beroda** [comment]
@akaszynski Sorry about the delay, I have been very busy lately.

I need to handle the conflicts and add a few comments, but I don't think I will implement more things. So you can start reviewing it whenever you want :smile:
Thank you again.

File: pyvista/core/dataset.py [hunk 1]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+
+        Concrete implementation of vtkDataSet's ``GetCellNeighbors`` method
+        (see https://vtk.org/doc/nightly/html/classvtkDataSet.html#ae1ba413c15802ef50d9b1955a66521e4)
+
+        Parameters
+        ----------
+        ind : int
+            Cell ID.
+
+        connections : str, optional
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2737]:
```suggestion
        connections : str, default: "points"
```

File: pyvista/core/dataset.py [hunk 2]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+            Describe how the neighbor cell(s) must be connected to the current
+            cell to be considered as a neighbor.
+            Can be either ``'points'``, ``'edges'`` or ``'faces'``.
+            Defaults to ``'points'``.
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2741]:
```suggestion
```

File: pyvista/core/dataset.py [hunk 3]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+
+        Returns
+        -------
+        List[int]
+            List of neighbor cells IDs for the ind-th cell.
+
+        Warnings
+        --------
+        For an ExplicitStructuredGrid, use the :func:`pyvista.ExplicitStructuredGrid.neighbors`
+        method.
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2751]:
```suggestion
        For an :class:`pyvista.ExplicitStructuredGrid`, use :func:`pyvista.ExplicitStructuredGrid.neighbors`.
```

File: pyvista/core/dataset.py [hunk 4]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+        --------
+        pyvista.DataSet.cell_neighbors_levels
+
+        Examples
+        --------
+        >>> from pyvista import examples
+        >>> mesh = examples.load_airplane()
+
+        Get the neighbor cell ids that have at least one point in common with
+        the 0-th cell
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2763]:
```suggestion
        the 0-th cell.
```

File: pyvista/core/dataset.py [hunk 5]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+
+        >>> mesh.cell_neighbors(0, "points")
+        [1, 2, 3, 388, 389, 11, 12, 395, 14, 209, 211, 212]
+
+        Get the neighbor cell ids that have at least one edge in common with
+        the 0-th cell
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2769]:
```suggestion
        the 0-th cell.
```

File: pyvista/core/dataset.py [hunk 6]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+
+        >>> mesh.cell_neighbors(0, "edges")
+        [1, 3, 12]
+
+        For unstructured grids with cells of dimension 3 (Tetrahedron for example),
+        cell neighbors can be defined using faces
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2775]:
```suggestion
        cell neighbors can be defined using faces.
```

File: pyvista/core/dataset.py [hunk 7]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+
+        >>> mesh = examples.download_tetrahedron()
+        >>> mesh.cell_neighbors(0, "faces")
+        [1, 5, 7]
+
+        Show a visual example
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2781]:
```suggestion
        Show a visual example.
```

File: pyvista/core/dataset.py [hunk 8]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+        List[int]
+            List of neighbor points IDs for the ind-th point.
+
+        See Also
+        --------
+        pyvista.DataSet.point_neighbors_levels
+
+        Examples
+        --------
+        Get the point neighbors of the 0-th point
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2893]:
```suggestion
        Get the point neighbors of the 0-th point.
```

File: pyvista/core/dataset.py [hunk 9]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+
+        >>> import pyvista as pv
+        >>> mesh = pv.Sphere(theta_resolution=10)
+        >>> mesh.point_neighbors(0)
+        [2, 226, 198, 170, 142, 114, 86, 254, 58, 30]
+
+        Plot them
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2900]:
```suggestion
        Plot them.
```

File: pyvista/core/dataset.py [hunk 10]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+        self, ind: int, n_levels: int = 1
+    ) -> Generator[List[int], None, None]:
+        """Get consecutive levels of point neighbors.
+
+        Parameters
+        ----------
+        ind : int
+            Point ID.
+
+        n_levels : int
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2941]:
```suggestion
        n_levels : int, default: 1
```

File: pyvista/core/dataset.py [hunk 11]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+            Number of levels to search for point neighbors.
+            When equals to 1, it is equivalent to pyvista.DataSet.point_neighbors.
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2943]:
```suggestion
            When equal to 1, it is equivalent to :func:`pyvista.DataSet.point_neighbors`.
```

File: pyvista/core/dataset.py [hunk 12]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+            A generator of list of neighbor points IDs for the ind-th point.
+
+        See Also
+        --------
+        pyvista.DataSet.point_neighbors
+
+        Examples
+        --------
+        Get the point neighbors IDs starting from the 0-th point
+        up until the third level
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2957]:
```suggestion
        up until the third level.
```

File: pyvista/core/dataset.py [hunk 13]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+        >>> import pyvista as pv
+        >>> mesh = pv.Sphere(theta_resolution=10)
+        >>> for neighbors in mesh.point_neighbors_levels(0, 3):
+        ...     print(neighbors)
+        ...
+        [2, 226, 198, 170, 142, 114, 86, 30, 58, 254]
+        [3, 227, 255, 199, 171, 143, 115, 87, 59, 31]
+        [256, 32, 4, 228, 200, 172, 144, 116, 88, 60]
+
+        Visualize these points IDs
```
Comment (akaszynski) [inline pyvista/core/dataset.py:2968]:
```suggestion
        Visualize these points IDs.
```

File: pyvista/core/dataset.py [hunk 14]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+        self, ind: int, connections: str = "points", n_levels: int = 1
+    ) -> Generator[List[int], None, None]:
+        """Get consecutive levels of cell neighbors.
+
+        Parameters
+        ----------
+        ind : int
+            Cell ID.
+
+        connections : str, optional
```
Comment (akaszynski) [inline pyvista/core/dataset.py:3014]:
```suggestion
        connections : str, default: "points"
```

File: pyvista/core/dataset.py [hunk 15]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+            Describe how the neighbor cell(s) must be connected to the current
+            cell to be considered as a neighbor.
+            Can be either ``'points'``, ``'edges'`` or ``'faces'``.
+            Defaults to ``'points'``.
```
Comment (akaszynski) [inline pyvista/core/dataset.py:3018]:
```suggestion
```

File: pyvista/core/dataset.py [hunk 16]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+
+        n_levels : int
+            Number of levels to search for cell neighbors.
+            When equals to 1, it is equivalent to pyvista.DataSet.cell_neighbors.
```
Comment (akaszynski) [inline pyvista/core/dataset.py:3022]:
```suggestion
            When equal to 1, it is equivalent to :func:`pyvista.DataSet.point_neighbors`.
```

File: pyvista/core/dataset.py [hunk 17]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+        ind : int
+            Cell ID.
+
+        connections : str, optional
+            Describe how the neighbor cell(s) must be connected to the current
+            cell to be considered as a neighbor.
+            Can be either ``'points'``, ``'edges'`` or ``'faces'``.
+            Defaults to ``'points'``.
+
+        n_levels : int
```
Comment (akaszynski) [inline pyvista/core/dataset.py:3020]:
```suggestion
        n_levels : int, default: 1
```

File: pyvista/core/dataset.py [hunk 18]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+
+        Returns
+        -------
+        List[int]
+            A generator of list of cell IDs for each level.
+
+        Warnings
+        --------
+        For an ExplicitStructuredGrid, use the :func:`pyvista.ExplicitStructuredGrid.neighbors`
+        method.
```
Comment (akaszynski) [inline pyvista/core/dataset.py:3032]:
```suggestion
        For an :class:`pyvista.ExplicitStructuredGrid`, use :func:`pyvista.ExplicitStructuredGrid.neighbors`.
```

File: pyvista/core/dataset.py [hunk 19]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+
+        See Also
+        --------
+        pyvista.DataSet.cell_neighbors
+
+        Examples
+        --------
+        Get the cell neighbors IDs starting from the 0-th cell
+        up until the third level
```
Comment (akaszynski) [inline pyvista/core/dataset.py:3041]:
```suggestion
        up until the third level.
```

File: pyvista/core/dataset.py [hunk 20]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+        >>>
+        >>> for ids in mesh.cell_neighbors_levels(
+        ...     0, connections="edges", n_levels=3
+        ... ):
+        ...     print(ids)
+        [1, 21, 9]
+        [2, 8, 74, 75, 20, 507]
+        [128, 129, 3, 453, 7, 77, 23, 506]
+
+        Visualize these cells IDs
```
Comment (akaszynski) [inline pyvista/core/dataset.py:3054]:
```suggestion
        Visualize these cells IDs.
```

File: pyvista/core/dataset.py [hunk 21]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+            Point ID.
+
+        Returns
+        -------
+        List[int]
+            List of cell IDs using the ind-th point.
+
+        Examples
+        --------
+        Get the cell ids using the 0-th point
```
Comment (akaszynski) [inline pyvista/core/dataset.py:3145]:
```suggestion
        Get the cell ids using the 0-th point.
```

File: pyvista/core/dataset.py [hunk 22]
```diff
@@ -2708,6 +2723,475 @@ def cell_point_ids(self, ind: int) -> List[int]:
...
+
+        >>> import pyvista as pv
+        >>> mesh = pv.Sphere(theta_resolution=10)
+        >>> mesh.point_cell_ids(0)
+        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+        Plot them
```
Comment (akaszynski) [inline pyvista/core/dataset.py:3152]:
```suggestion
        Plot them.
```

File: pyvista/core/filters/data_set.py [hunk 23]
```diff
@@ -4336,14 +4336,17 @@ def plot_over_circular_arc_normal(
...
-    def extract_cells(self, ind, progress_bar=False):
+    def extract_cells(self, ind, invert=False, progress_bar=False):
         """Return a subset of the grid.
 
         Parameters
         ----------
         ind : numpy.ndarray
             Numpy array of cell indices to be extracted.
 
+        invert : bool
```
Comment (akaszynski) [inline pyvista/core/filters/data_set.py:4347]:
```suggestion
        invert : bool, default: False
```

**Comment by akaszynski** [comment]
Excellent examples!

image

**Review by akaszynski — APPROVED** [review]
Made several changes:
- Applied `blackdoc`. I'm now 100% convinced we need this; it saved me so much time having to patch up the example sections.
- Removed `print` from examples. We avoid printing in order to make sure our documentation generation is clean from `stdout`.
- Fixed links. We like to use links (internal/external) whenever possible. You did a great job with most of these, I just added more.
- Using `generator` in the docs since we're returning a generator.

Overall I'm really impressed with this and quite sold on including this within our API. I would have preferred that the typing changes not directly associated with this feature remain out of this PR, but I'll let it slide. I'm guilty of going overboard and this is a sub 1k LOC PR.

**Comment by beroda** [comment]
@akaszynski Thank you for your feedback, it is very much appreciated !

> I would have preferred that the typing changes not directly associated with this feature remain out of this PR, but I'll let it slide. I'm guilty of going overboard and this is a sub 1k LOC PR.

Actually, typing changes have been introduced in this PR by the pre-commit hook `pyupgrade`, see 
https://github.com/asottile/pyupgrade#remove-quoted-annotations
It is activated when you use `from __future__ import annotations`

**Review by banesullivan — APPROVED** [review]
Thanks so much for this contribution! The implementation, examples, and tests all look excellent!

I have some minor performance concerns with the for-loops and comprehensions, but these can be optimized later if needed or used more heavily

**Pull request merged**